### PR TITLE
Add option to sanity check tombstones on reads/compaction

### PR DESCRIPTION
**Pull request by aweisberg**

### CODE REVIEW

File: src/java/org/apache/cassandra/db/UnfilteredValidation.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+import org.apache.cassandra.config.Config;
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.db.rows.Unfiltered;
+
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.serializers.MarshalException;
+import org.apache.cassandra.utils.NoSpamLogger;
+
+public class UnfilteredValidation
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/db/UnfilteredValidation.java]: Can you add a brief high level class description about what this  class attempts to validate?

File: src/java/org/apache/cassandra/db/UnfilteredValidation.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        String msg = String.format("Key %s in %s.%s is invalid in %s: %s",
+                                   metadata.partitionKeyType.getString(key.getKey()),
+                                   metadata.keyspace,
+                                   metadata.name,
+                                   filename,
+                                   invalidContent);
+        if (strat == Config.CorruptedTombstoneStrategy.exception)
+            throw new MarshalException(msg);
+        else if (strat == Config.CorruptedTombstoneStrategy.warn)
+            nospam1m.warn(msg);
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/db/UnfilteredValidation.java]: Warn or error? These are errors right?

**Comment by aweisberg** [comment]
Should we have this validation enabled for utests and dtests and maybe have the dtests fail if these validations fail?

File: test/unit/org/apache/cassandra/db/compaction/CompactionsCQLTest.java [hunk 3]
```diff
@@ -245,6 +256,174 @@ public void testPerCFSNeverPurgeTombstones() throws Throwable
         testPerCFSNeverPurgeTombstonesHelper(false);
     }
 
+    @Test
+    public void testCompactionInvalidRTs() throws Throwable
+    {
+        DatabaseDescriptor.setCorruptedTombstoneStrategy(Config.CorruptedTombstoneStrategy.exception);
+        prepare();
+        RangeTombstone rt = new RangeTombstone(Slice.ALL, new DeletionTime(System.currentTimeMillis(), -1));
```
Comment (aweisberg) [inline test/unit/org/apache/cassandra/db/compaction/CompactionsCQLTest.java]: For each test case can you explain exactly what field you are setting that is invalid and then ELI5 reason why that value isn't valid.

**Review by aweisberg — COMMENTED** [review]
In addition to the can we run this in the dtests and utests comment.

My only concern is that maybe this additional validation in some weird case might trigger an NPE because it's validating something that is null some of the time. I reviewed them and I think there are no cases of that, but it's worth you giving it a look to see if you spot any.

File: src/java/org/apache/cassandra/db/UnfilteredValidation.java (hunk 2 above)
Comment (krummas) [reply src/java/org/apache/cassandra/db/UnfilteredValidation.java]: My thinking here was that if the operator consider these to be errors, they would enable exceptions, if not they just think they are warnings and then the log level should be warn, I have no strong feelings about this though

File: test/unit/org/apache/cassandra/cql3/validation/operations/TTLTest.java [hunk 4]
```diff
@@ -167,6 +169,8 @@ public void testRejectExpirationOverflowPolicy() throws Throwable
     @Test
     public void testRecoverOverflowedExpirationWithScrub() throws Throwable
     {
+        // this tests writes corrupt tombstones on purpose, disable the strategy:
+        DatabaseDescriptor.setCorruptedTombstoneStrategy(Config.CorruptedTombstoneStrategy.disabled);
```
Comment (aweisberg) [inline test/unit/org/apache/cassandra/cql3/validation/operations/TTLTest.java:187]: Should you enable it again afterwards?

File: test/unit/org/apache/cassandra/db/compaction/CompactionsCQLTest.java [hunk 5]
```diff
@@ -425,11 +441,21 @@ private void readAndValidate(boolean asc) throws Throwable
...
+            assertNotNull(cause);
+            MarshalException me = (MarshalException) cause;
+            assertTrue(me.getMessage().contains("Key 22"));
         }
         assertTrue(gotException);
+        for (SSTableReader sstable : getCurrentColumnFamilyStore().getLiveSSTables())
+        {
+            assertTrue(sstable.isMarkedSuspect());
+            sstable.unmarkSuspect();
+        }
```
Comment (aweisberg) [inline test/unit/org/apache/cassandra/db/compaction/CompactionsCQLTest.java]: All of these things are copy pasted three different tests. Could be a little more DRY but it's not a big deal.

File: src/java/org/apache/cassandra/db/UnfilteredValidation.java (hunk 2 above)
Comment (aweisberg) [reply src/java/org/apache/cassandra/db/UnfilteredValidation.java]: Makes sense.

File: src/java/org/apache/cassandra/db/UnfilteredValidation.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        if (strat == Config.CorruptedTombstoneStrategy.exception)
+        {
+            // we mark suspect to make sure this sstable is not included in future compactions - it would just keep
+            // throwing exceptions
+            sstable.markSuspect();
+            throw new CorruptSSTableException(new MarshalException(msg), sstable.getFilename());
+        }
+        else if (strat == Config.CorruptedTombstoneStrategy.warn)
+        {
+            nospam1m.warn(msg);
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/db/UnfilteredValidation.java]: NoSpamLogger uses the string you pass in as a key for the timer on the 1 minute limit. So this is not going to ratelimit if you are have a lot of bad keys/invalidContent. I think you want to format the keyspace, name, and sstable, but not invalidContent and key string. Then format in the keyString and invalidContent as params to NoSpamLogger.

Another thing you can do is that it returns a boolean indicating whether it logged so you can count how many were suppressed, but that's non-trivial plumbing since you would need to have a map of counters for each keyspace, cf, and sstable.

I think ponies might be that you enhance NoSpamLogger so you can retrieve a log statement by using the generated key (log string) and it contains the suppressed log message count so you can just grab it and format it in. NoSpamLogger can then zero it out when it successfully logs.

File: src/java/org/apache/cassandra/db/UnfilteredValidation.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        if (strat != Config.CorruptedTombstoneStrategy.disabled && unfiltered != null && !unfiltered.isEmpty())
+        {
+            boolean hasInvalidDeletions = false;
+            try
+            {
+                hasInvalidDeletions = unfiltered.hasInvalidDeletions();
+            }
+            catch (Throwable t) // make sure no unknown exceptions fail the read/compaction
+            {
+                nospam1m.error("Could not check if Unfiltered in {} had any invalid deletions", sstable, t);
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/db/UnfilteredValidation.java:59]: The exception at the end won't get formatted in. NoSpamLogger only supports the varargs signature and it can't tell if you have an exception at the end.

Comment (krummas) [reply src/java/org/apache/cassandra/db/UnfilteredValidation.java:59]: looks like NoSpamLogger just passes the objects down to the underlying logger, and with slf4j the stack trace will be printed if there is an exception as the last item in that array: https://www.slf4j.org/faq.html#paramException

File: test/unit/org/apache/cassandra/cql3/validation/operations/TTLTest.java [hunk 8]
```diff
@@ -170,10 +170,12 @@ public void testRejectExpirationOverflowPolicy() throws Throwable
     public void testRecoverOverflowedExpirationWithScrub() throws Throwable
     {
         // this tests writes corrupt tombstones on purpose, disable the strategy:
+        Config.CorruptedTombstoneStrategy original = DatabaseDescriptor.getCorruptedTombstoneStrategy();
         DatabaseDescriptor.setCorruptedTombstoneStrategy(Config.CorruptedTombstoneStrategy.disabled);
         baseTestRecoverOverflowedExpiration(false, false);
         baseTestRecoverOverflowedExpiration(true, false);
         baseTestRecoverOverflowedExpiration(true, true);
+        DatabaseDescriptor.setCorruptedTombstoneStrategy(original);
```
Comment (aweisberg) [inline test/unit/org/apache/cassandra/cql3/validation/operations/TTLTest.java]: If the test fails it will stay disabled. 

Maybe store the original in before test and set it back in after test? That way whether or not it changes it will go back to the original value?

**Comment by aweisberg** [comment]
Believe this is no longer in use.
